### PR TITLE
KAFKA-16518; Adding standalone argument for storage 

### PR DESCRIPTION
**Pull request by muralibasani**
Resolves : https://issues.apache.org/jira/browse/KAFKA-16518

- Adds a new argument "standalone" to kafka-storage.sh
- If standalone mode, creates a checkpoint file in metadata dir ${kafkaConfig.metadataLogDir}/__cluster_metadata-0/

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by muralibasani** [comment]
@jsancio what do you think on this approach ?
it's abit not clear on creating the kRaftVersionRecord and votersRecord though.

```

./bin/kafka-storage.sh format --cluster-id oS80nAuvTOyliP131PWSFb --config ./config/kraft/server.properties --standalone
SLF4J: Class path contains multiple SLF4J bindings.
SLF4J: Found binding in [jar:file:/Users/muralidharbasani/projects/others/kafka3/kafka/tools/build/dependant-libs-2.13.14/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: Found binding in [jar:file:/Users/muralidharbasani/projects/others/kafka3/kafka/trogdor/build/dependant-libs-2.13.14/slf4j-reload4j-1.7.36.jar!/org/slf4j/impl/StaticLoggerBinder.class]
SLF4J: See http://www.slf4j.org/codes.html#multiple_bindings for an explanation.
SLF4J: Actual binding is of type [org.slf4j.impl.Reload4jLoggerFactory]
metaPropertiesEnsemble=MetaPropertiesEnsemble(metadataLogDir=Optional.empty, dirs={/tmp/kraft-combined-logs: EMPTY})
logDir : /tmp/kraft-combined-logs
Formatting /tmp/kraft-combined-logs with metadata.version 3.7-IV4.
bootstrapDirectory : org.apache.kafka.metadata.bootstrap.BootstrapDirectory@1fb19a0
Snapshot written to /tmp/kraft-combined-logs/__cluster_metadata-0
muralidharbasani@Muralidhars-MacBook-Pro kafka % cat /tmp/kraft-combined-logs/__cluster_metadata-0/00000000000000000000-0000000000.checkpoint 
WYJ�� ������������������*
V�Ѣ�������������������$����"����?hY� �����������������%    
```

File: core/src/main/scala/kafka/tools/StorageTool.scala [hunk 1]
```diff
@@ -452,21 +460,68 @@ object StorageTool extends Logging {
...
+
+    // Create the full path for the checkpoint file
+    val checkpointFilePath = snapshotDir.resolve(snapshotDir)
+
+    // Create the raw snapshot writer
+    val rawSnapshotWriter = FileRawSnapshotWriter.create(checkpointFilePath, new OffsetAndEpoch(0, 0))
+
+    val builder = new RecordsSnapshotWriter.Builder()
+      .setKraftVersion(1)
+      .setVoterSet(Optional.empty())
```
Comment (jsancio) [inline core/src/main/scala/kafka/tools/StorageTool.scala]: If you create and set the `VoterSet` to contain this one replica, The builder will write the necessary `KRaftVersionRecord` and `VotersRecord`. To create a voter set you need the default listener name (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/raft/RaftManager.scala#L188), the replica id (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/raft/RaftManager.scala#L222), the directory id (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/SharedServer.scala#L260) and the advertise hostname and port (see https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/server/KafkaConfig.scala#L1237-L1243).

Looking at this code (https://github.com/apache/kafka/commit/e8d32563f323a245b3127394d97d6356e84f0294#diff-cbe6a8b71b05ed22cf09d97591225b588e9fca6caaf95d3b34a43262cfd23aa6R2190-R2193), we are going to have to relax this to allowed advertise listeners as long as `controller.quorum.voters` is not set.

File: core/src/main/scala/kafka/tools/StorageTool.scala [hunk 2]
```diff
@@ -452,21 +460,68 @@ object StorageTool extends Logging {
...
+      .setRawSnapshotWriter(rawSnapshotWriter).build(new MetadataRecordSerde)
+
+    // Append control records: KRaftVersionRecord and VotersRecord
+    appendControlRecords(builder)
```
Comment (jsancio) [inline core/src/main/scala/kafka/tools/StorageTool.scala]: This actually write those records as "metadata" records and not as controller records.

You don't need to call this method as the `RecordsSnapshotWriter.Builder.build(RecordSerde)` will write the necessary control record if you configure the builder correctly by calling `setKraftVersion` and `setVoterSet`.

**Review by jsancio — COMMENTED** [review]
@muralibasani thanks for the feature. I left some comments.

File: core/src/main/scala/kafka/tools/StorageTool.scala (hunk 1 above)
Comment (muralibasani) [reply core/src/main/scala/kafka/tools/StorageTool.scala]: @jsancio Thanks. when trying to create a voterset, VoterNode constructor is not accessible in core\tools, as VoterNode is part of raft package. Any suggestions ?

We need this public constructor https://github.com/apache/kafka/blob/8068a086a3f41b9f7e4d4a1dab3338f029089f23/raft/src/main/java/org/apache/kafka/raft/internals/VoterSet.java#L49

Comment (muralibasani) [reply core/src/main/scala/kafka/tools/StorageTool.scala]: trying this way 
```
    
  val voterMap: java.util.Map[Integer, VoterSet.VoterNode] = new util.HashMap()
      voterMap.put(1, new VoterSet.VoterNode(
        ReplicaKey.of(nodeId, Optional.of(directoryId)),
        Collections.singletonMap(controllerListenerNames, new InetSocketAddress(host, port)),
        new SupportedVersionRange(0.toShort, 1.toShort)
      ))
  val voterSet = new VoterSet(voterMap)

```

**Pull request closed without merging**